### PR TITLE
Cherry pick GameBoy-Online updates

### PR DESCRIPTION
**Pull request by s3ththompson**
I just cherry-picked two updates that @taisel has made to the original GameBoy-Online `GameBoyCore.js`.

### CODE REVIEW

**Comment by s3ththompson** [comment]
Great project btw :+1:

**Comment by jbmorley** [comment]
@s3ththompson Thanks so much for the PR (and the donation!); I'll get this into a release over the next few days. What's does it fix specifically for the release notes?

**Comment by taisel** [comment]
There's a lot more that needs to be done to the original project...

**Comment by taisel** [comment]
At this point though, a rewrite would just be a c++ compilation to asm.js

**Comment by taisel** [comment]
Thirdly, the original audio backend does not contain workarounds for audio on iOS safari. iOS safari suspends web audio API until a function of it is called within an touchend event. The original audio backend attempts to create a context outside of such as soon as possible. The XAudioJS still needs a clean way to integrate that workaround.

**Comment by taisel** [comment]
Also, the original licensing was GPL2, but I changed IodineGBA to BSD type and probably should relicense the gb/gbc emulator to BSD as well.

**Comment by jbmorley** [comment]
@taisel Great to have some input – thanks!

I'm currently looking to move to a submodule model for GameBoy-Online and also integrate IodineGBA as well: hopefully that would mean that I'd be able to keep up with any updates and make it easier to contribute upstream. This would also mean that I'd pick up any licensing changes automatically.

In terms of licensing – if you're keen to move to a BSD-type license for these emulators, there's nothing stopping me switching gameplay over (other than the original GameBoy-Online license which I necessarily inherited).

The audio issues you're describing would perhaps explain some of the weirdness I've been seeing lately on the iPhone 6 (though I'm not yet convinced this isn't an iOS bug), but I've not really had much opportunity to investigate further.

As for separating out the UI, I'm obviously happy to help in crafting APIs which help make this easier as the interface between the gameplay UI code (far from elegant in of itself) and GameBoy-Online is murky at best.

**Comment by taisel** [comment]
Yeah... the interface code between the emulator and GUI is a total shitshow in the original gameboy-online project. Lessons were learned and gui interfacing *should* be easier to work with IodineGBA. Be careful though, that GBA emulator pulls its own voodoo with webworker API shimming. ;)

File: src/gbo/GameBoyCore.js [hunk 1]
```diff
@@ -4986,7 +4986,7 @@ GameBoyCore.prototype.recomputeDimension = function () {
 	//Cache some dimension info:
 	// this.onscreenWidth = this.canvas.width;
 	// this.onscreenHeight = this.canvas.height;
-	if (window && window.mozRequestAnimationFrame) {
+	if (window && window.mozRequestAnimationFrame || (navigator.userAgent.toLowerCase().indexOf("gecko") != -1 && navigator.userAgent.toLowerCase().indexOf("like gecko") == -1)) {
```
Comment (taisel) [inline src/gbo/GameBoyCore.js]: This is a shitshow riiiiight here. Firefox gets canvas scaled by CSS and webkit/blink based browsers get canvas scaled by context2d. The reason being Firefox doesn't hw accelerate context2d on the platform I test on while webkit/blink does, while at the same time Firefox accelerates scaling via CSS when webkit/blink do not.

Newer Firefox removed the deprecated mozRequestAnimationFrame we were sniffing on, so it defaulted to the webkit/blink friendly path when it shouldn't in Firefox 4X.

Comment (taisel) [inline src/gbo/GameBoyCore.js]: Deep down in blink/webkit someone decided to not accelerate image scaling via CSS if the width/height were below 256 pixels in dimension.

Comment (taisel) [inline src/gbo/GameBoyCore.js]: Feel free to insert a better test here if possible.

File: src/gbo/GameBoyCore.js [hunk 2]
```diff
@@ -7895,12 +7895,20 @@ GameBoyCore.prototype.memoryReadJumpCompile = function () {
...
 					}
 					break;
-				case 0xFF76:
-				case 0xFF77:
-					this.memoryHighReader[index & 0xFF] = this.memoryReader[index] = function (parentObj, address) {
-						return 0;
-					}
-					break;
+                case 0xFF76:
+                    //Undocumented realtime PCM amplitude readback:
```
Comment (taisel) [inline src/gbo/GameBoyCore.js]: These exist only in GBC mode. I think it was implemented wrong, in that we're supposed to be reading back the value after the duty cycle masking out the volume, not simply the volume.

Comment (taisel) [inline src/gbo/GameBoyCore.js]: The whole GB/GBC emulator core is a shitshow that should be discontinued for a rewritten one. There's a lot of not obvious structural timing issues that requires a lot of thought put in to track down. @_@

Comment (taisel) [inline src/gbo/GameBoyCore.js]: In other words... +1 on asm.js redo.

Comment (taisel) [inline src/gbo/GameBoyCore.js]: I need to fix regs 0x76 and x077 and the HDMA loop reading out of bounds in a rare case. I need to reopen issue tracking. :I

**Comment by taisel** [comment]
GameBoy-Online might be less buggy than IodineGBA, but IodineGBA sure as hell is much more maintainable.

**Comment by s3ththompson** [comment]
Thanks for chiming in @taisel! I just cherry-picked the updates in the name of project health.

Would be happy to help contribute if you reopened GameBoy-Online bug tracking.

Agreed that an asm.js rewrite makes sense in the long run--but not sure how it would affect iOS Safari perf before they implement true asm.js AOT support. (Also there's some value in having a readable pure JS emulator for pedagogy and debuggability)

**Pull request merged**